### PR TITLE
Fix emitted intermediate state for initial incremental non-CDC syncs

### LINKED ISSUE #24808: Erroneous state associated with unsuccessful initial non-CDC incremental sync
There is currently a bug with the initial full refresh sync while performing a non-CDC incremental sync. 

The logic for emitting intermediate state for incremental syncs currently depends on the fact that the records are returned by the iterator in order. In fact, it's disabled for certain sources (BigQuery and MongoDB) due their inability to order records from an incremental sync by the cursor field. 

In all of our JDBC sources, the way an initial full refresh sync is performed is by a ``select * from table``. Subsequent syncs do order by the cursor field. Crucially, the results are never ordered by the cursor field. As a result, the emitted intermediate and final state is wrong. 

Consider the following table (with a configured intermediate state emission frequency of 1)
| ID(cursor)  | name |
| ------------- | ------------- |
| 5  | Duy  |
| 1  | Subodh  |
| 3  | Akash  |
| 2  | Rodi  |
| 4  | Prateek  |

With the given ``select *`` query, we're not guaranteed to get the results in order. This can be problematic if there is an unsuccessful sync, which can lead to data loss.
e.g. If the source connector processes row 1 and fails, it will have emitted one state which is ``{ cursor_value : 5 (Duy) }``. Subsequent syncs will skip all values with ``cursor_value < 5`` with no path to recovery. This may explain some of the data loss for incr syncs. However, it would only correspond to data in the initial load.  

Some of the reasons we may not be seeing more of these issues is : 
1. Sync completed successfully. 
2. The cursor is indexed and somehow the `select *` query returned everything in order.

The fix here is to : 
1. Order the initial load by cursor field. One potential drawback here is that if the source database is not indexed by cursor field, sorting could take a long time. 
2.  Change the state emitting logic to only return one state at the end for the initial load (which is the max value seen so far). The drawback of this approach is that we won't have any real checkpointing during the initial load - arguably when we need it the most. 

Am inclined to go with option 1

### PR DESCRIPTION
**Pull request by akashkulk**
Closes https://github.com/airbytehq/airbyte/issues/24808

An initial incremental sync now process records in the order of their cursor value for JDBC sources. Please see the parent issue for a more detailed write-up of the issue. 

Main changes: 
1. `AbstractJdbcSource.queryTableFullRefresh()` : Issues a different query based on whether the originating sync was a `FULL_REFRESH` vs `INCREMENTAL` sync. For the latter case, we sort the results by the cursor value now. 
2. `StateDecoratingIterator` : Adds logging in case the records are NOT emitted in order from the underlying stream. This is mainly for debugging purposes. Note that failing the stream here would likely lead to data loss
3. `PostgresSourceTest` : Adds a unit test that (i) changes intermediate state emission frequency to 1 (ii) verifies that records and state follow our invariants. (iii) tests the next sync

### CODE REVIEW

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java [hunk 1]
```diff
@@ -101,11 +103,22 @@ public AbstractJdbcSource(final String driverClass,
...
     LOGGER.info("Queueing query for table: {}", tableName);
-    return queryTable(database, String.format("SELECT %s FROM %s",
-        enquoteIdentifierList(columnNames, getQuoteString()),
-        getFullyQualifiedTableNameWithQuoting(schemaName, tableName, getQuoteString())));
+    // This corresponds to the initial sync for in INCREMENTAL_MODE. The ordering of the records matters as intermediate state messages are emitted.
+    if (syncMode.equals(SyncMode.INCREMENTAL)) {
+      final String quotedCursorField = enquoteIdentifier(cursorField.get(), getQuoteString());
+      return queryTable(database, String.format("SELECT %s FROM %s ORDER BY %s ASC",
+          enquoteIdentifierList(columnNames, getQuoteString()),
+          getFullyQualifiedTableNameWithQuoting(schemaName, tableName, getQuoteString()), quotedCursorField));
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java:115]: `quotedCursorField` 
Checking: Is this going to be the correct fully qualified name with correct escaping?

Comment (akashkulk) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java:115]: Yup - I've added an additional log in `queryTable`. An example for a sample table is `SELECT \"id\",\"name\",\"power\" FROM \"public\".\"id_and_name\" ORDER BY \"id\" ASC"`

It's the same logic that is used in the existing incremental query (where we do sort by cursor) : https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java#L305

**Review by rodireich — APPROVED** [review]

**Review by nguyenaiden — COMMENTED** [review]
Few questions, but looks good so far!

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java [hunk 2]
```diff
@@ -101,11 +103,22 @@ public AbstractJdbcSource(final String driverClass,
...
+                                                                  final String tableName,
+                                                                  final SyncMode syncMode,
+                                                                  final Optional<String> cursorField) {
     LOGGER.info("Queueing query for table: {}", tableName);
-    return queryTable(database, String.format("SELECT %s FROM %s",
-        enquoteIdentifierList(columnNames, getQuoteString()),
-        getFullyQualifiedTableNameWithQuoting(schemaName, tableName, getQuoteString())));
+    // This corresponds to the initial sync for in INCREMENTAL_MODE. The ordering of the records matters as intermediate state messages are emitted.
+    if (syncMode.equals(SyncMode.INCREMENTAL)) {
+      final String quotedCursorField = enquoteIdentifier(cursorField.get(), getQuoteString());
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java:112]: It should crash in that case. Ideally we shouldn't get to a point with an incremental sync w/out a cursor defined.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java [hunk 3]
```diff
@@ -94,15 +95,28 @@ public static Source sshWrappedSource() {
...
+                                                               final String tableName,
+                                                               final SyncMode syncMode,
+                                                               final Optional<String> cursorField) {
     LOGGER.info("Queueing query for table: {}", tableName);
+    // This corresponds to the initial sync for in INCREMENTAL_MODE. The ordering of the records matters as intermediate state messages are emitted.
+    if (syncMode.equals(SyncMode.INCREMENTAL)) {
+      final String quotedCursorField = enquoteIdentifier(cursorField.get(), getQuoteString());
+      final String newIdentifiers = getWrappedColumnNames(database, null, columnNames, schemaName, tableName);
+      final String preparedSqlQuery =
+          String.format("SELECT %s FROM %s ORDER BY %s ASC", newIdentifiers,
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:107]: Ah fair point, I'm going to tackle some of that in the refactoring work coming up. The other thing that I want to check is why and if we can get rid of the override in MsSQL's case

**Review by nguyenaiden — APPROVED** [review]
Good for me! +1

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (nguyenaiden) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:107]: I'm good with that. Go for it!

**Comment by akashkulk** [comment]
/publish connector=connectors/source-postgres run-tests=false

**Comment by akashkulk** [comment]
/publish connector=connectors/source-postgres-strict-encrypt run-tests=false

**Comment by akashkulk** [comment]
/publish connector=connectors/source-mysql-strict-encrypt run-tests=false

**Comment by akashkulk** [comment]
/publish connector=connectors/source-mssql-strict-encrypt run-tests=false

**Comment by akashkulk** [comment]
/publish connector=connectors/source-alloydb-strict-encrypt run-tests=false

**Pull request merged**